# Data Wrangling 

In [1]:
import pandas as pd

## Loading

In [236]:
bcd = pd.read_csv('./data/breast_cancer_data.csv')
bcd

patient_id  clump_thickness  cell_size_uniformity  cell_shape_uniformity  \
0       1000025              5.0                   1.0                      1   
1       1002945              5.0                   4.0                      4   
2       1015425              3.0                   1.0                      1   
3       1016277              6.0                   8.0                      8   
4       1017023              4.0                   1.0                      1   
..          ...              ...                   ...                    ...   
694      776715              3.0                   1.0                      1   
695      841769              2.0                   1.0                      1   
696      888820              5.0                  10.0                     10   
697      897471              4.0                   8.0                      6   
698      897471              4.0                   8.0                      8   

     marginal_adhesion  single_ep_cell_size bare_nuclei  bland_chromatin  \
0                    1                    2           1              3.0   
1                    5                    7          10              3.0   
2                    1                    2           2              3.0   
3                    1                    3           4              3.0   
4                    3                    2           1              3.0   
..                 ...                  ...         ...              ...   
694                  1                    3           2              1.0   
695                  1                    2           1              1.0   
696                  3                    7           3              8.0   
697                  4                    3           4             10.0   
698                  5                    4           5             10.0   

     normal_nucleoli  mitoses      class doctor_name  
0                1.0        1     benign     Dr. Doe  
1                2.0        1     benign   Dr. Smith  
2                1.0        1     benign     Dr. Lee  
3                7.0        1     benign   Dr. Smith  
4                1.0        1     benign    Dr. Wong  
..               ...      ...        ...         ...  
694              1.0        1     benign     Dr. Lee  
695              1.0        1     benign   Dr. Smith  
696             10.0        2  malignant     Dr. Lee  
697              6.0        1  malignant     Dr. Lee  
698              4.0        1  malignant    Dr. Wong  

[699 rows x 12 columns]

## Exploring the Data Frame

In [64]:
bcd.describe()

patient_id  clump_thickness  cell_size_uniformity  \
count  6.990000e+02       698.000000            698.000000   
mean   1.071704e+06         4.416905              3.137536   
std    6.170957e+05         2.817673              3.052575   
min    6.163400e+04         1.000000              1.000000   
25%    8.706885e+05         2.000000              1.000000   
50%    1.171710e+06         4.000000              1.000000   
75%    1.238298e+06         6.000000              5.000000   
max    1.345435e+07        10.000000             10.000000   

       cell_shape_uniformity  marginal_adhesion  single_ep_cell_size  \
count             699.000000         699.000000           699.000000   
mean                3.207439           2.793991             3.216023   
std                 2.971913           2.843163             2.214300   
min                 1.000000           1.000000             1.000000   
25%                 1.000000           1.000000             2.000000   
50%                 1.000000           1.000000             2.000000   
75%                 5.000000           3.500000             4.000000   
max                10.000000          10.000000            10.000000   

       bland_chromatin  normal_nucleoli     mitoses  
count       695.000000       698.000000  699.000000  
mean          3.447482         2.868195    1.589413  
std           2.441191         3.055647    1.715078  
min           1.000000         1.000000    1.000000  
25%           2.000000         1.000000    1.000000  
50%           3.000000         1.000000    1.000000  
75%           5.000000         4.000000    1.000000  
max          10.000000        10.000000   10.000000

In [65]:
bcd.shape

(699, 12)

In [66]:
bcd.dtypes

patient_id                 int64
clump_thickness          float64
cell_size_uniformity     float64
cell_shape_uniformity      int64
marginal_adhesion          int64
single_ep_cell_size        int64
bare_nuclei               object
bland_chromatin          float64
normal_nucleoli          float64
mitoses                    int64
class                     object
doctor_name               object
dtype: object

In [237]:
bcd.nunique()

patient_id               645
clump_thickness           10
cell_size_uniformity      10
cell_shape_uniformity     10
marginal_adhesion         10
single_ep_cell_size       10
bare_nuclei               11
bland_chromatin           10
normal_nucleoli           10
mitoses                    9
class                      2
doctor_name                4
dtype: int64

In [68]:
bcd.columns

Index(['patient_id', 'clump_thickness', 'cell_size_uniformity',
       'cell_shape_uniformity', 'marginal_adhesion', 'single_ep_cell_size',
       'bare_nuclei', 'bland_chromatin', 'normal_nucleoli', 'mitoses', 'class',
       'doctor_name'],
      dtype='object')

## Modifying a text column

Let's look at the column for the doctors' names:

In [99]:
bcd['doctor_name']

0        Dr. Doe
1      Dr. Smith
2        Dr. Lee
3      Dr. Smith
4       Dr. Wong
         ...    
694      Dr. Lee
695    Dr. Smith
696      Dr. Lee
697      Dr. Lee
698     Dr. Wong
Name: doctor_name, Length: 699, dtype: object

The doctors' name data are redundant; each one has a "Dr. " in front of the actual name, but we already know these are doctors by the column name. And the entries have white space in them, which is just annoying. So let's modify this column so it only contains the actual surnames of the doctors.

One great thing about pandas is that it has versions of many of Python's string methods that operate element-wise on column of strings. We want to separate the "Dr. " from the actual name, which is exactly what Python's `str.split()` function does. So chances are, pandas has a version of this function that operates element-wise on data frames.

---

#### String Splitting Review:

Let's briefly remind ourselves of splitting up Python strings and extracting bits of them.

In [70]:
# Here's a string of the form: surname, first initial.
myStr = 'SirString, A.'
print(myStr)

SirString, A.


Let's say we wanted to get the surname. We could split this string into a list at the white space like this:

In [71]:
spltStr = myStr.split()
print(spltStr)

['SirString,', 'A.']


We now have a list in which the items contain the text on either side of the split. This is close: the first entry in the list has the surname, but it also has an unwanted comma. 

Let's split the string at the comma instead:

In [72]:
spltStr = myStr.split(',')   # tell Python to split at commas
print(spltStr)

['SirString', ' A.']


And now we can get the surname by indexing:

In [73]:
surname = spltStr[0]
print(surname)

SirString


---

See if you can extract the surname from `myStr` in one line of code:

In [74]:
surname = myStr.split(',')[0]
print(surname)

SirString


---

Alright, time to replace the `bcd[doctor_name]` column values with just the doctors' surnames. 

We could do this in one step, but let's break it out for clarity. First, let's copy the name column out into a new series.

In [238]:
dr_names = bcd['doctor_name']
dr_names

0        Dr. Doe
1      Dr. Smith
2        Dr. Lee
3      Dr. Smith
4       Dr. Wong
         ...    
694      Dr. Lee
695    Dr. Smith
696      Dr. Lee
697      Dr. Lee
698     Dr. Wong
Name: doctor_name, Length: 699, dtype: object

Now let's split all the names in the `doctor_name` column at the whitespace by using pandas `DataFrame.str.split()` function.

In [239]:
split_dr_names = dr_names.str.split()
split_dr_names

0        [Dr., Doe]
1      [Dr., Smith]
2        [Dr., Lee]
3      [Dr., Smith]
4       [Dr., Wong]
           ...     
694      [Dr., Lee]
695    [Dr., Smith]
696      [Dr., Lee]
697      [Dr., Lee]
698     [Dr., Wong]
Name: doctor_name, Length: 699, dtype: object

Now we have a column of lists, each with two elements. The first element of each list is the "Dr. " bit, and the second consists of the surnames we want. 

We can get these by using pandas string indexing, `Series.str[index]`. 

In [240]:
surnames = split_dr_names.str[1]
surnames[0] = 'hello'
surnames

0      hello
1      Smith
2        Lee
3      Smith
4       Wong
       ...  
694      Lee
695    Smith
696      Lee
697      Lee
698     Wong
Name: doctor_name, Length: 699, dtype: object

In [242]:
bcd['doctor_name'] = surnames

In [243]:
bcd['doctor_name']

0      hello
1      Smith
2        Lee
3      Smith
4       Wong
       ...  
694      Lee
695    Smith
696      Lee
697      Lee
698     Wong
Name: doctor_name, Length: 699, dtype: object

Success!

## Converting a column type (and other aggravations)

In [223]:
bcd.dtypes

patient_id                 int64
clump_thickness          float64
cell_size_uniformity     float64
cell_shape_uniformity      int64
marginal_adhesion          int64
single_ep_cell_size        int64
bare_nuclei               object
bland_chromatin          float64
normal_nucleoli          float64
mitoses                    int64
class                     object
doctor_name               object
dtype: object

In [224]:
bcd['bare_nuclei'] = bcd['bare_nuclei'].astype('int64')

ValueError: invalid literal for int() with base 10: '?'

And, argh, we get an error! If we look at the bottom of the error message, it looks like the error involves question marks in the data, which would also explain why this column imported as text rather than numbers in the first place.

Let's check.

In [244]:
bcd[bcd['bare_nuclei'] == '?']

patient_id  clump_thickness  cell_size_uniformity  cell_shape_uniformity  \
23      1057013              8.0                   4.0                      5   
40      1096800              6.0                   6.0                      6   
139     1183246              1.0                   1.0                      1   
145     1184840              1.0                   1.0                      3   
158     1193683              1.0                   1.0                      2   
164     1197510              5.0                   1.0                      1   
235     1241232              3.0                   1.0                      4   
249      169356              3.0                   1.0                      1   
275      432809              3.0                   1.0                      3   
292      563649              8.0                   8.0                      8   
294      606140              1.0                   1.0                      1   
297       61634              5.0                   4.0                      3   
315      704168              4.0                   6.0                      5   
321      733639              3.0                   1.0                      1   
411     1238464              1.0                   1.0                      1   
617     1057067              1.0                   1.0                      1   

     marginal_adhesion  single_ep_cell_size bare_nuclei  bland_chromatin  \
23                   1                    2           ?              7.0   
40                   9                    6           ?              7.0   
139                  1                    1           ?              2.0   
145                  1                    2           ?              2.0   
158                  1                    3           ?              1.0   
164                  1                    2           ?              3.0   
235                  1                    2           ?              3.0   
249                  1                    2           ?              3.0   
275                  1                    2           ?              2.0   
292                  1                    2           ?              6.0   
294                  1                    2           ?              2.0   
297                  1                    2           ?              2.0   
315                  6                    7           ?              4.0   
321                  1                    2           ?              3.0   
411                  1                    1           ?              2.0   
617                  1                    1           ?              1.0   

     normal_nucleoli  mitoses      class doctor_name  
23               3.0        1  malignant       Smith  
40               8.0        1     benign        Wong  
139              1.0        1     benign         Lee  
145              1.0        1     benign        Wong  
158              1.0        1     benign         Doe  
164              1.0        1     benign       Smith  
235              1.0        1     benign         Doe  
249              1.0        1     benign        Wong  
275              1.0        1     benign         Lee  
292             10.0        1  malignant        Wong  
294              1.0        1     benign         Lee  
297              3.0        1     benign       Smith  
315              9.0        1     benign        Wong  
321              1.0        1     benign       Smith  
411              1.0        1     benign         Lee  
617              1.0        1     benign       Smith

Sure enough. Rather than leaving the cells of missing values empty, somebody has made the poor decision to enter question marks.

When you are dealing with other peoples' data, you'll find that this sort of the happens a LOT. It can be very aggravating.

Let's replace the question marks with nothing, so that this column becomes consistent with the rest.

In [245]:
bcd['bare_nuclei'] = bcd['bare_nuclei'].replace('?', '')

And now we can convert the column to numeric values.

In [246]:
bcd['bare_nuclei'] = pd.to_numeric(bcd['bare_nuclei'])

In [247]:
bcd.dtypes

patient_id                 int64
clump_thickness          float64
cell_size_uniformity     float64
cell_shape_uniformity      int64
marginal_adhesion          int64
single_ep_cell_size        int64
bare_nuclei              float64
bland_chromatin          float64
normal_nucleoli          float64
mitoses                    int64
class                     object
doctor_name               object
dtype: object

Okay! We have now have gotten our data somewhat into shape, meaning:

- missing data are actually missing
- columns of numeric data are numeric in type
- the column of doctor names contains only last names

So now we can explore some ways to deal with missing values.

## Dealing with missing data

### Finding missing values

Even though this data set isn't all that large:

In [248]:
bcd.shape

(699, 12)

699 rows is lot to look through in order to find missing values by hand. 

We can test for missing values using the `DataFrame.isna()` method. 

In [249]:
bcd.isna()

patient_id  clump_thickness  cell_size_uniformity  cell_shape_uniformity  \
0         False            False                 False                  False   
1         False            False                 False                  False   
2         False            False                 False                  False   
3         False            False                 False                  False   
4         False            False                 False                  False   
..          ...              ...                   ...                    ...   
694       False            False                 False                  False   
695       False            False                 False                  False   
696       False            False                 False                  False   
697       False            False                 False                  False   
698       False            False                 False                  False   

     marginal_adhesion  single_ep_cell_size  bare_nuclei  bland_chromatin  \
0                False                False        False            False   
1                False                False        False            False   
2                False                False        False            False   
3                False                False        False            False   
4                False                False        False            False   
..                 ...                  ...          ...              ...   
694              False                False        False            False   
695              False                False        False            False   
696              False                False        False            False   
697              False                False        False            False   
698              False                False        False            False   

     normal_nucleoli  mitoses  class  doctor_name  
0              False    False  False        False  
1              False    False  False        False  
2              False    False  False        False  
3              False    False  False        False  
4              False    False  False        False  
..               ...      ...    ...          ...  
694            False    False  False        False  
695            False    False  False        False  
696            False    False  False        False  
697            False    False  False        False  
698            False    False  False        False  

[699 rows x 12 columns]

By itself, that doesn't help us much. But if we combine it with summation (remember that `True` values count as 1 and `False` counts as zero):

In [250]:
bcd.isna().sum()

patient_id                0
clump_thickness           1
cell_size_uniformity      1
cell_shape_uniformity     0
marginal_adhesion         0
single_ep_cell_size       0
bare_nuclei              18
bland_chromatin           4
normal_nucleoli           1
mitoses                   0
class                     0
doctor_name               0
dtype: int64

Now we have the counts by variable, and can easly see that there are missing values for a few of the variables.

The bare nuclei variable we dealt with earlier has the most missing values, with bland chromatin coming a distant second. 

Let's check some of the rows with missing values and make sure everything else looks normal in those rows.

In [251]:
bcd[bcd['bland_chromatin'].isna()]

patient_id  clump_thickness  cell_size_uniformity  cell_shape_uniformity  \
342      814265              2.0                   1.0                      1   
343      814911              1.0                   1.0                      1   
359      873549             10.0                   3.0                      5   
365      897172              2.0                   1.0                      1   

     marginal_adhesion  single_ep_cell_size  bare_nuclei  bland_chromatin  \
342                  1                    2          1.0              NaN   
343                  1                    2          1.0              NaN   
359                  4                    3          7.0              NaN   
365                  1                    2          1.0              NaN   

     normal_nucleoli  mitoses      class doctor_name  
342              1.0        1     benign         Lee  
343              1.0        1     benign         Doe  
359              5.0        3  malignant         Doe  
365              1.0        1     benign         Lee

---

In the cell below, check the rows that have missing values for clump thickness and cell size uniformity. Do this in one go rather than separately.

In [252]:
bcd[bcd['clump_thickness'].isna() | bcd['cell_size_uniformity'].isna()]

patient_id  clump_thickness  cell_size_uniformity  cell_shape_uniformity  \
6      1018099              1.0                   NaN                      1   
12     1041801              NaN                   3.0                      3   

    marginal_adhesion  single_ep_cell_size  bare_nuclei  bland_chromatin  \
6                   1                    2         10.0              3.0   
12                  3                    2          3.0              4.0   

    normal_nucleoli  mitoses      class doctor_name  
6               1.0        1     benign         Doe  
12              4.0        1  malignant       Smith

---

So far so good. It looks like the rows that have missing values just have one missing value, and everything else seems fine. But let's do check that no rows have more than one missing value.

To do this, we can sum the number of missing values across the columns (i.e. for each row), and then see what the maximum number of missing values within a row is.

In [253]:
row_na_totals = bcd.isna().sum(axis = 1)
row_na_totals.max()

1

---

In the cell below, do the above calculation in one line.

In [254]:
bcd.isna().sum(axis = 1).max()

1

### Dealing with missing values

Now that we have determined that there are missing values, we have to determine how to deal with them. 

#### Ignoring missing values elementwise

One way to handle missing values is just to ignore them. Most of the standard math and statisitical functions will do that by default.

So this:

In [255]:
bcd['clump_thickness'].mean()  

4.416905444126074

Computes the mean clump thickness ignoring the one missing value.

We can compute the mean (again ignoring missing values) for all the numeric columns like this:

In [257]:
bcd.mean(numeric_only = True)  # the numeric_only refers to columns, not missing values

patient_id               1.071704e+06
clump_thickness          4.416905e+00
cell_size_uniformity     3.137536e+00
cell_shape_uniformity    3.207439e+00
marginal_adhesion        2.793991e+00
single_ep_cell_size      3.216023e+00
bare_nuclei              3.538913e+00
bland_chromatin          3.447482e+00
normal_nucleoli          2.868195e+00
mitoses                  1.589413e+00
dtype: float64

That worked, but the output is a little awkward because the patient ID is being treated as a numeric variable. We can fix that by converting the patient ID variable to a string variable.

In [258]:
bcd['patient_id'] = bcd['patient_id'].astype('string')

And now the means should look a little better because we won't have the mean for the ID column in the millions>

In [260]:
bcd.mean(numeric_only = True)

clump_thickness          4.416905
cell_size_uniformity     3.137536
cell_shape_uniformity    3.207439
marginal_adhesion        2.793991
single_ep_cell_size      3.216023
bare_nuclei              3.538913
bland_chromatin          3.447482
normal_nucleoli          2.868195
mitoses                  1.589413
dtype: float64

#### Removing missing values

Before we start messing around too much with the values in our data frame, let's make sure we can easily hit the reset button and get back to a nice starting point. To do this, we'll want to

- reload the data
- modify the column of Dr. names
- set the patient ID to type str 
- set the bare nuclei column to numeric

This is a perfect job for a function!


---

In the cell below, write function to reset our data frame to the desired starting point.

In [261]:
def hit_reset():
    bcd = pd.read_csv('./data/breast_cancer_data.csv')
    bcd['patient_id'] = bcd['patient_id'].astype('string')
    bcd['doctor_name'] = bcd['doctor_name'].str.split().str[1]
    bcd['bare_nuclei'] = bcd['bare_nuclei'].replace('?', '')
    bcd['bare_nuclei'] = pd.to_numeric(bcd['bare_nuclei'])
    
    return bcd

---

##### Removing rows with missing values

Obviously, rows in which all values are missing won't do us any good, so we can drop them with:

In [145]:
bcd = bcd.dropna(how = 'all')

This drops rows in which *all* of the values are missing. This code ran without error, but we know it also didn't do anything in this case because we don't have any rows in which all the values are missing.

Sometimes a case can be made for throwing out all observations (rows) that are incomplete, that is, if they contain *any* missing values.

In [262]:
bcd = bcd.dropna(how = 'any')

---

In the cell below, check the (new) shape of `bcd`.

In [263]:
bcd.shape

(674, 12)

It should have fewer rows now.

---

And now is a perfect time to test our function! In the cell below, hit the reset button on bcd.

In [264]:
bcd = hit_reset()  # reset bcd data frame

Check the shape.

In [265]:
bcd.shape

(699, 12)

Check the data types of the columns.

In [266]:
bcd.dtypes

patient_id                string
clump_thickness          float64
cell_size_uniformity     float64
cell_shape_uniformity      int64
marginal_adhesion          int64
single_ep_cell_size        int64
bare_nuclei              float64
bland_chromatin          float64
normal_nucleoli          float64
mitoses                    int64
class                     object
doctor_name               object
dtype: object

Check the doctor name column.

In [267]:
bcd['doctor_name']

0        Doe
1      Smith
2        Lee
3      Smith
4       Wong
       ...  
694      Lee
695    Smith
696      Lee
697      Lee
698     Wong
Name: doctor_name, Length: 699, dtype: object

---

##### Removing columns with missing values

And we could do the same for columns if we wished, though this is less frequently done. We just need to change the axis (direction) over which `DataFrame.dropna()` works.

In [271]:
bcd = bcd.dropna(axis = 1, how = 'any') # drop columns rather than rows

This leaves us with only the complete columns.

In [269]:
bcd.shape

(699, 7)

Let's see which they are.

In [270]:
bcd.columns

Index(['patient_id', 'cell_shape_uniformity', 'marginal_adhesion',
       'single_ep_cell_size', 'mitoses', 'class', 'doctor_name'],
      dtype='object')

#### Filling in missing values

Occasionally, we may want to fill in missing values. This isn't very common, but might be useful if some other function you are using doesn't handle missing values gracefully.

Before filling in missing values, we need to restore our data frame so it actually has missing values. Good thing we wrote that function!

In [272]:
bcd = hit_reset()

We can fill in missing values with any single value we want, such as a zero.

In [184]:
bcd = bcd.fillna(0)

---

In the cell below, check to see that we no longer have missing values.

In [273]:
bcd.isna().sum()

patient_id                0
clump_thickness           1
cell_size_uniformity      1
cell_shape_uniformity     0
marginal_adhesion         0
single_ep_cell_size       0
bare_nuclei              18
bland_chromatin           4
normal_nucleoli           1
mitoses                   0
class                     0
doctor_name               0
dtype: int64

In the cell below, reset the data and verify that the missing data are back.

In [274]:
bcd = hit_reset()
bcd.isna().sum()

patient_id                0
clump_thickness           1
cell_size_uniformity      1
cell_shape_uniformity     0
marginal_adhesion         0
single_ep_cell_size       0
bare_nuclei              18
bland_chromatin           4
normal_nucleoli           1
mitoses                   0
class                     0
doctor_name               0
dtype: int64

---

In the cell below, fill the missing values in each column with the column mean. (Hint: this is pandas, so this is actually easy!)

In [235]:
bcd = bcd.fillna(bcd.mean(numeric_only = True))

patient_id               0
clump_thickness          0
cell_size_uniformity     0
cell_shape_uniformity    0
marginal_adhesion        0
single_ep_cell_size      0
bare_nuclei              0
bland_chromatin          0
normal_nucleoli          0
mitoses                  0
class                    0
doctor_name              0
dtype: int64

And now verify that there are no more missing values.

In [275]:
bcd.isna().sum()

patient_id                0
clump_thickness           1
cell_size_uniformity      1
cell_shape_uniformity     0
marginal_adhesion         0
single_ep_cell_size       0
bare_nuclei              18
bland_chromatin           4
normal_nucleoli           1
mitoses                   0
class                     0
doctor_name               0
dtype: int64

---